File: Data Structures & Algorithms/binary-tree-right-side-view/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        
        right_side_nodes = [] # stores right most nodes' values
        if root is None: return right_side_nodes # guard rail for when tree has 0 nodes

        nodes = deque([root]) # start bfs with something in there

        while nodes:

            len_of_level = len(nodes)

            for i in range(len_of_level):
                
                node = nodes.popleft()

                if i == len_of_level - 1: # last thing in the deque
                    right_side_nodes.append(node.val)
                
                if node.left:
                    nodes.append(node.left)
                if node.right:    
                    nodes.append(node.right)
        
        return right_side_nodes

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        return count_good_nodes(root, root->val);
    }


    int count_good_nodes(TreeNode* root, int curr_max){
        if (!root) return 0;

        auto res = (root->val >= curr_max ? 1 : 0);
        curr_max = std::max(curr_max, root->val);
        res += count_good_nodes(root->left, curr_max);
        res += count_good_nodes(root->right, curr_max);
        
        return res;

    }
};
